libraries/discriminator/src/discriminator.rs

```diff
@@ -0,0 +1,72 @@
+//! The traits and types used to create a discriminator for a type
+
+use {
+    bytemuck::{Pod, Zeroable},
+    solana_program::{hash, program_error::ProgramError},
+};
+
+/// A trait for managing 8-byte discriminators in a slab of bytes
+pub trait SplDiscriminates {
+    /// The 8-byte discriminator as a `[u8; 8]`
+    const SPL_DISCRIMINATOR: ArrayDiscriminator;
+    /// The 8-byte discriminator as a slice (`&[u8]`)
+    const SPL_DISCRIMINATOR_SLICE: &'static [u8] = Self::SPL_DISCRIMINATOR.as_slice();
+}
+
+/// Array Discriminator type
+#[derive(Clone, Copy, Debug, Default, PartialEq, Pod, Zeroable)]
+#[repr(transparent)]
+pub struct ArrayDiscriminator([u8; ArrayDiscriminator::LENGTH]);
+impl ArrayDiscriminator {
+    /// Size for discriminator in account data
+    pub const LENGTH: usize = 8;
+    /// Uninitialized variant of a discriminator
+    pub const UNINITIALIZED: Self = Self::new([0; Self::LENGTH]);
+    /// Creates a discriminator from an array
+    pub const fn new(value: [u8; Self::LENGTH]) -> Self {
+        Self(value)
+    }
+    /// Get the array as a const slice
+    pub const fn as_slice(&self) -> &[u8] {
+        self.0.as_slice()
+    }
+    /// Creates a new `ArrayDiscriminator` from some hash input string literal
+    pub fn new_with_hash_input(hash_input: &str) -> Self {
+        let hash_bytes = hash::hashv(&[hash_input.as_bytes()]).to_bytes();
+        let mut discriminator_bytes = [0u8; 8];
+        discriminator_bytes.copy_from_slice(&hash_bytes[..8]);
+        Self(discriminator_bytes)
+    }
+    /// Returns the discriminator length using the common `len()` method
+    pub fn len(&self) -> usize {
```

Comment:
Maybe I'm missing something, but why is this needed? It's an array, so it doesn't really have a length

Reply:
Just for the common nomenclature of calling `len()` on it instead of `ArrayDiscriminator::LENGTH`, and we chatted about opening this up for variable discriminator lengths, so this sets the stage for that.

If you think it's unnecessary I can remove

Reply:
Yeah sure! Assuming `len()` is more common pattern for things with variable size that can grow/shrink, like vectors? 
Our "variable length" discriminator is still fixed once initialized

Reply:
Done deal, removed